dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/integration/RegistryDirectory.java

```diff
@@ -187,37 +177,38 @@ public void destroy() {
 
     @Override
     public synchronized void notify(List<URL> urls) {
-        List<URL> categoryUrls = urls.stream().filter(this::isValidCategory).filter(this::isNotCompatibleFor26x).collect(Collectors.toList());
+        List<URL> categoryUrls = urls.stream()
+                .filter(this::isValidCategory)
+                .filter(this::isNotCompatibleFor26x)
+                .collect(Collectors.toList());
 
         /**
          * TODO Try to refactor the processing of these three type of urls using Collectors.groupBy()?
          */
-        this.configurators = Configurator.toConfigurators(classifyUrls(categoryUrls, url -> (CONFIGURATORS_CATEGORY.equals(url.getParameter(Constants.CATEGORY_KEY, Constants.DEFAULT_CATEGORY))
-                 || OVERRIDE_PROTOCOL.equals(url.getProtocol())))).orElse(configurators);
+        this.configurators = Configurator.toConfigurators(classifyUrls(categoryUrls, UrlUtils::isConfigure))
+                .orElse(configurators);
 
-        toRouters(classifyUrls(categoryUrls, url -> {
-            return ROUTE_PROTOCOL.equals(url.getProtocol())
-                    || ROUTERS_CATEGORY.equals(url.getParameter(Constants.CATEGORY_KEY, Constants.DEFAULT_CATEGORY));
-        })).ifPresent(this::addRouters);
+        toRouters(classifyUrls(categoryUrls, UrlUtils::isRoute)).ifPresent(this::addRouters);

```

Comment:
If I understood correctly then here, you are transforming urls to routes using 'toRouters' method and if there if any of the urls got transformed to route object then adding them to **routerChain** using `ifPresent(this::addRouters)`.

I looked at **toRouters**  code and found there it is already getting added to **routerChain** . 

Based on these analysis do you see is there any chance of having duplicate route information in routerChain?

I might be thinking a loud, so please correct me.

Reply:
you are correct. This is a flaw, I will fix it.

Reply:
@beiwei30 are you planning to consider this fix as part of this PR or planning another PR?